tidal-music/tidal-sdk-android
# Set up API client code generation

This PR adds the main code generation setup to `:tidalapi`, as well as the generated files.

Please review commit by commit to have an easy overview, and just skip the [commit adding the generated files](https://github.com/tidal-music/tidal-sdk-android/commit/0aff75b49815716cc8f5f36fc4e056e12649dee2) 😄 

## Note for reviewers:
The code in this pull request is a MVP. Please point out anything you don't like, but the idea is to get this one merged and then improve what needs improving in subsequent PRs. So if you request changes it should be for things that absolutely **can't** go in like this.
Please also look at the additional info provided in the commit messages.

## Note about the generated code:
I debated whether it should be in the `src` folder or not, but after some reading I decided yes, because:
- We will commit the generated code
- Generation will not happen as part of the normal build process
- Generated code in the future will be changed and reviewed via PR.

## How can you test this?
1. Check out the branch
2. Publish to `MavenLocal`
3. Add `MavenLocal` in `settings.gradle.kts` like this:
```
dependencyResolutionManagement {
    repositories {
        mavenCentral()
        }
}
```
4. Add a dependency to `com.tidal.sdk.tidalapi:0.1.0` where you want to use it, and instantiate an api client like this:
```
 val client: TidalApiClient = TidalApiClient(credentialsProvider)
 ```
 5. Play around with it.

## Do you need to test it to review this?
At this stage, no imho. But if you like to, feel free!

## How to/ test code generation?
As you currently need a binary of my fork of `openapi-generator`, ask me personally, if you want to play around with it.
But after this PR, we wil quickly move towards making this available via GH Actions, which will simplify this all.
